app/decorators/order_decorator.rb

```diff
@@ -15,4 +15,26 @@ def state_badge
       helpers.tag.span('failed', class: 'order-state badge badge-danger')
     end
   end
+
+  def events_buttons
+    permitted_events = Orders::StateService.new(object).permitted_events
+    %w[complete fail].each do |event|
+      h.concat method("#{event}_button").call if permitted_events.include?(event.to_sym)
```

Comment:
WDT about move permitted events to a method and use memoization there?

Reply:
done -  also for payments and shipments for the sake of consistency